reposerver/repository/utils.go

```diff
@@ -0,0 +1,55 @@
+package repository
+
+import (
+	"path/filepath"
+	"strings"
+
+	"github.com/argoproj/argo-cd/v2/reposerver/apiclient"
+)
+
+// GetApplicationRootPath returns the common root path among a set of application-related paths for manifest generation.
+func GetApplicationRootPath(q *apiclient.ManifestRequest, appPath, repoPath string) string {
+	paths := getPaths(q, appPath, repoPath)
+
+	if len(paths) == 0 {
+		// backward compatibility, by default the root path is the repoPath
+		return repoPath
+	}
+
+	commonParts := strings.Split(paths[0], "/")
+
+	for _, path := range paths[1:] {
+		parts := strings.Split(path, "/")
+		// Find the minimum length between the current common parts and the current path
+		minLen := func(a, b int) int {
+			if a < b {
+				return a
+			}
+			return b
+		}(len(commonParts), len(parts))
+
+		for i := 0; i < minLen; i++ {
+			if commonParts[i] != parts[i] {
+				commonParts = commonParts[:i]
+				break
+			}
+		}
+	}
+	return strings.Join(commonParts, "/")
+}
+
+// getPaths retrieves all absolute paths associated with the generation of application manifests.
+func getPaths(q *apiclient.ManifestRequest, appPath, repoPath string) []string {
+	var paths []string
+	for _, annotationPath := range strings.Split(q.AnnotationManifestGeneratePaths, ";") {
+		if annotationPath == "" {
+			continue
+		}
+		if filepath.IsAbs(annotationPath) {
+			paths = append(paths, filepath.Clean(filepath.Join(repoPath, annotationPath)))
+		} else {
+			paths = append(paths, filepath.Clean(filepath.Join(appPath, annotationPath)))
```

Comment:
`filepath.Clean` doesn't evaluate symlinks, so it's not really safe when handling user input. Can you replace `filepath.Clean`/`filepath.Join` with SecureJoin?

Reply:
Hi! replacing `filepath.Clean/filepath.Join` by `SecureJoin` breaks unit tests cause the result is not the same:
`SecureJoin` doesn't support relatives paths like `../../`

```go
// SecureJoinVFS joins the two given path components (similar to Join) except
// that the returned path is guaranteed to be scoped inside the provided root
// path (when evaluated).
```

Then,  for an application with appPath: `/tmp/_argocd-repo/7a58c52a-0030-4fd9-8cc5-35b2d8b4e731/services/helloworld`  with an annotation of `../../overlays` instead to return `/tmp/_argocd-repo/7a58c52a-0030-4fd9-8cc5-35b2d8b4e731/overlays` it is returning `/tmp/_argocd-repo/7a58c52a-0030-4fd9-8cc5-35b2d8b4e731/services/helloworld/overlays`


Any idea how to fix it? Thanks

Reply:
Ah sure.

So, all we care about is preventing the user from escaping the repo root. We can safely use `filepath.Clean` on any path that's relative to the repo root, and then securejoin that relative path to the repo root.

Maybe something like this:

```go
		if filepath.IsAbs(annotationPath) {
			paths = append(paths, securejoin.SecureJoin(repoPath, filepath.Clean(annotationPath)))
		} else {
			relativeAppPath := somehowGetRelativeAppPath(appPath)
			paths = append(paths, securejoin.SecureJoin(repoPath, filepath.Clean(filepath.Join(relativeAppPath, annotationPath))))
		}
```

Reply:
Ohh thanks! Done